autoload/lsp/internal/linked_editing_range.vim

```diff
@@ -0,0 +1,165 @@
+    \         lsp#callbag#subscribe({
+    \           'next': { x -> call('s:prepare', x) }
+    \         })
+    \     ),
+    \     lsp#callbag#pipe(
+    \         lsp#callbag#fromEvent(['InsertLeave']),
+    \         lsp#callbag#filter({ -> g:lsp_linked_editing_range_enabled }),
+    \         lsp#callbag#subscribe({ -> s:clear() })
+    \     ),
+    \     lsp#callbag#pipe(
+    \         lsp#callbag#fromEvent(['TextChanged', 'TextChangedI', 'TextChangedP']),
+    \         lsp#callbag#filter({ -> g:lsp_linked_editing_range_enabled }),
+    \         lsp#callbag#subscribe({ -> s:sync() })
+    \     ),
+    \ )
+endfunction
+
+function! lsp#internal#linked_editing_range#_disable() abort
+    if exists('s:Dispose')
+        call s:clear()
+        call s:Dispose()
+        unlet s:Dispose
+    endif
+endfunction
+
+function! lsp#internal#linked_editing_range#prepare() abort
+    if !s:enabled()
+        return ''
+    endif
+
+    call lsp#callbag#pipe(
+    \     s:request_sync(),
+    \     lsp#callbag#subscribe({
+    \         'next': { x -> call('s:prepare', x) },
+    \         'error': { -> {} },
+    \     })
+    \ )
+    return ''
+endfunction
+
+function! s:enabled(...) abort
+    return g:lsp_linked_editing_range_enabled && s:TextMark.is_available()
+endfunction
+
+function! s:request_sync() abort
+    let l:server = lsp#get_allowed_servers(&filetype)
+    let l:server = filter(l:server, 'lsp#capabilities#has_linked_editing_range_provider(v:val)')
+    let l:server = get(l:server, 0, v:null)
+    if empty(l:server)
+        return lsp#callbag#of([v:null])
+    endif
+
+    return lsp#callbag#of(
+    \     lsp#callbag#pipe(
+    \         lsp#request(l:server, {
+    \             'method': 'textDocument/linkedEditingRange',
+    \             'params': {
+    \                 'textDocument': lsp#get_text_document_identifier(),
+    \                 'position': lsp#get_position(),
+    \             }
+    \         }),
+    \         lsp#callbag#toList()
+    \     ).wait({ 'sleep': 1, 'timeout': 200 })
+    \ )
+endfunction
+
+function! s:prepare(x) abort
+    if empty(a:x) || empty(get(a:x, 'response')) || empty(get(a:x['response'], 'result')) || empty(get(a:x['response']['result'], 'ranges'))
+        return
+    endif
+
+    call s:clear()
+    call s:TextMark.set(bufnr('%'), s:TEXT_MARK_NAMESPACE, map(a:x['response']['result']['ranges'], { _, range -> {
+    \     'start_pos': lsp#utils#position#lsp_to_vim('%', range['start']),
+    \     'end_pos': lsp#utils#position#lsp_to_vim('%', range['end']),
+    \     'highlight': 'Underlined',
+    \ } }))
+    let s:state['bufnr'] = bufnr('%')
+    let s:state['changenr'] = changenr()
+    let s:state['changedtick'] = b:changedtick
+endfunction
+
+function! s:clear() abort
+    call s:TextMark.clear(bufnr('%'), s:TEXT_MARK_NAMESPACE)
+endfunction
+
+function! s:sync() abort
+    let l:bufnr = bufnr('%')
+    if s:state['bufnr'] != l:bufnr
```

Comment:
for callbags this is usually not preferred. you can instead use `switchMap` and `takeUntil` similar to this https://github.com/prabirshrestha/vim-lsp/blob/master/autoload/lsp/internal/document_highlight.vim

Feel free to change or ignore it for now and I can refactor it later when it is merged.

I usually use RxJS or other Rx libraries does to understand callbag.

High level these are what it means.
1. switchMap -> when ever you have a next, cancel the existing one and start a new one. Good example for this is autocomplete where if I type `he` and then type `hello` I want to cancel `he` and automatically start `hello`.
2. takeUntil ->  this is use for cancellation. basically it means when a next is called in takeuntil cancel the stream. 


But since you are using wait what you have should be ok.

Some good reading here. https://medium.com/@benlesh/rxjs-dont-unsubscribe-6753ed4fda87

Reply:
Fixed.

Reply:
Thank you. I learned a lot.